cmdk/src/index.tsx

```diff
@@ -992,6 +1015,39 @@ const useScheduleLayoutEffect = () => {
   }
 }
 
+function renderChildren(children: React.ReactElement) {
+  const childrenType = children.type as any
+  // The children is a component
+  if (typeof childrenType === 'function') return childrenType(children.props)
+  // The children is a component with `forwardRef`
+  else if ('render' in childrenType) return childrenType.render(children.props)
+  // It's a string, boolean, etc.
+  else return children
+}
+
+function SlottableWithNestedChildren(
+  { asChild, children }: { asChild?: boolean; children?: React.ReactNode },
+  render: (child: React.ReactNode) => JSX.Element,
+) {
+  if (asChild && React.isValidElement(children)) {
+    return React.cloneElement(renderChildren(children), { ref: (children as any).ref }, render(children.props.children))
+  }
+  return render(children)
+}
+
+const Primitive = NODES.reduce((primitive, node) => {
+  const Node = React.forwardRef((props: PrimitivePropsWithRef<typeof node>, forwardedRef: any) => {
+    const { asChild, ...primitiveProps } = props
+    const Comp: any = asChild ? Slot : node
+
+    return <Comp {...primitiveProps} ref={forwardedRef} />
+  })
+
+  Node.displayName = `Primitive.${node}`
+
+  return { ...primitive, [node]: Node }
+}, {} as Primitives)
```

Comment:
Let's bring in the `@radix-ui/react-primitive` package for this behavior instead, for better package de-duplication.